File: apps/backend/src/schemas/budget.py
from datetime import datetime
from typing import Literal, Optional
from pydantic import BaseModel, Field


class BudgetBase(BaseModel):
    name: str
    category: str
    amount: float = Field(gt=0)
    currency: str = Field(default="PHP", max_length=3)
    period: Literal["weekly", "monthly", "yearly"]
    start_date: str
    color: str = "#7C6BFF"
    emoji: str = "💰"


class BudgetCreate(BudgetBase):
    pass


class BudgetUpdate(BaseModel):
    name: Optional[str] = None
    amount: Optional[float] = Field(default=None, gt=0)
    period: Optional[Literal["weekly", "monthly", "yearly"]] = None
    color: Optional[str] = None
    emoji: Optional[str] = None


class BudgetRead(BudgetBase):
    id: str
    user_id: str
    spent: float
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class BudgetListResponse(BaseModel):
    items: list[BudgetRead]
    total: int

File: apps/mobile/src/types/budget.ts
import type { SyncStatus } from './account';

export interface Budget {
  id: string;
  userId: string;
  name: string;
  category: string;
  amount: number;
  spent: number;
  currency: string;
  period: 'custom' | 'monthly' | 'yearly';
  startDate: string;
  endDate?: string;
  color: string;
  emoji: string;
  createdAt: string;
  updatedAt: string;
  syncStatus?: SyncStatus;
}

export interface CreateBudgetDto {
  name: string;
  category: string;
  amount: number;
  currency: string;
  period: Budget['period'];
  startDate: string;
  endDate?: string;
  color: string;
  emoji: string;
}

export interface BudgetWithProgress extends Budget {
  progressPercent: number;
  remaining: number;
  isOverBudget: boolean;
}

export interface BudgetGroup {
  name: string;
  period: 'custom' | 'monthly' | 'yearly';
  startDate: string;
  endDate?: string;
  color: string;
  emoji: string;
  totalAmount: number;
  totalSpent: number;
  categories: Budget[];
}

export interface CreateBudgetGroupDto {
  name: string;
  period: 'custom' | 'monthly' | 'yearly';
  startDate: string;
  endDate?: string;
  color: string;
  emoji: string;
  currency: string;
  categories: {
    categoryId: string;
    amount: number;
  }[];
}
